packages/flutter/lib/src/widgets/fade_in_image.dart

```diff
@@ -309,267 +333,176 @@ class FadeInImage extends StatefulWidget {
-      oldImageStream?.removeListener(listener);
-      _imageStream.addListener(listener);
-    }
-  }
-
-  void _handleImageChanged(ImageInfo imageInfo, bool synchronousCall) {
-    _imageInfo = imageInfo;
-    listener();
-  }
-
-  void stopListening() {
-    _imageStream?.removeListener(ImageStreamListener(_handleImageChanged));
-  }
-}
-
 class _FadeInImageState extends State<FadeInImage> with TickerProviderStateMixin {
-  _ImageProviderResolver _imageResolver;
-  _ImageProviderResolver _placeholderResolver;
-
   AnimationController _controller;
-  Animation<double> _animation;
-
-  FadeInImagePhase _phase = FadeInImagePhase.start;
-  FadeInImagePhase get phase => _phase;
+  Animation<double> _fadeOutAnimation;
+  Animation<double> _fadeInAnimation;
 
   @override
   void initState() {
-    _imageResolver = _ImageProviderResolver(state: this, listener: _updatePhase);
-    _placeholderResolver = _ImageProviderResolver(state: this, listener: () {
-      setState(() {
-        // Trigger rebuild to display the placeholder image
-      });
-    });
-    _controller = AnimationController(
-      value: 1.0,
-      vsync: this,
-    );
-    _controller.addListener(() {
-      setState(() {
-        // Trigger rebuild to update opacity value.
-      });
-    });
-    _controller.addStatusListener((AnimationStatus status) {
-      _updatePhase();
-    });
     super.initState();
+    _controller = AnimationController(vsync: this);
+    _initializeAnimations();
   }
 
   @override
-  void didChangeDependencies() {
-    _resolveImage();
-    super.didChangeDependencies();
+  void dispose() {
+    _fadeInAnimation.removeStatusListener(_handleFadeInAnimationStatusChange);
+    _fadeInAnimation.removeListener(_handleFadeInTick);
+    _controller.dispose();
+    super.dispose();
   }
 
   @override
   void didUpdateWidget(FadeInImage oldWidget) {
     super.didUpdateWidget(oldWidget);
-    if (widget.image != oldWidget.image || widget.placeholder != oldWidget.placeholder)
-      _resolveImage();
+    if (!_controller.isAnimating &&
+        (widget.fadeOutDuration != oldWidget.fadeOutDuration ||
+         widget.fadeInDuration != oldWidget.fadeInDuration ||
+         widget.fadeOutCurve != oldWidget.fadeOutCurve ||
+         widget.fadeInCurve != oldWidget.fadeInCurve)) {
+      _initializeAnimations();
+    }
+    if (widget.image != oldWidget.image) {
+      _controller.reset();
+    }
   }
 
-  @override
-  void reassemble() {
-    _resolveImage(); // in case the image cache was flushed
-    super.reassemble();
+  void _initializeAnimations() {
+    final double split = widget.fadeOutDuration.inMicroseconds /
+        (widget.fadeOutDuration + widget.fadeInDuration).inMicroseconds;
+    _fadeOutAnimation = Tween<double>(begin: 1, end: 0).animate(
+      CurvedAnimation(
+        parent: _controller,
+        curve: Interval(0, split, curve: widget.fadeOutCurve),
+      ),
+    );
+    _fadeInAnimation = Tween<double>(begin: 0, end: 1).animate(
```

Comment:
A `Tween<double>(begin: 0, end: 1)` is a no-op, so this is just wasted CPU cycles

Reply:
I reworked this implementation a bunch, so this code isn't relevant anymore - resolving.